File: bin/spice/pkg/display/table.go
/*
Copyright 2024-2025 The Spice.ai OSS Authors

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

     https://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package display

import (
	"fmt"
	"strings"
)

// Table displays data in a formatted table with optional column types
func Table(colNames []string, colTypes []string, rows [][]string, colWidths []int) {
	// Print top border
	fmt.Print("+")
	for _, width := range colWidths {
		fmt.Print(strings.Repeat("-", width+2))
		fmt.Print("+")
	}
	fmt.Println()

	// Print column names (centered)
	fmt.Print("|")
	for i, colName := range colNames {
		padding := colWidths[i] - len(colName)
		leftPad := padding / 2
		rightPad := padding - leftPad
		fmt.Printf(" %s%s%s |", strings.Repeat(" ", leftPad), colName, strings.Repeat(" ", rightPad))
	}
	fmt.Println()

	// Print column types if provided (centered)
	if colTypes != nil {
		fmt.Print("|")
		for i, colType := range colTypes {
			padding := colWidths[i] - len(colType)
			leftPad := padding / 2
			rightPad := padding - leftPad
			fmt.Printf(" %s%s%s |", strings.Repeat(" ", leftPad), colType, strings.Repeat(" ", rightPad))
		}
		fmt.Println()
	}

	// Print header separator
	fmt.Print("+")
	for _, width := range colWidths {
		fmt.Print(strings.Repeat("-", width+2))
		fmt.Print("+")
	}
	fmt.Println()

	// Print all rows
	for _, row := range rows {
		// Split each cell by newlines to handle multi-line content
		maxLines := 1
		cellLines := make([][]string, len(row))
		for col, value := range row {
			cellLines[col] = strings.Split(value, "\n")
			if len(cellLines[col]) > maxLines {
				maxLines = len(cellLines[col])
			}
		}

		// Print each line of the row
		for lineIdx := 0; lineIdx < maxLines; lineIdx++ {
			fmt.Print("|")
			for col := range row {
				var lineValue string
				if lineIdx < len(cellLines[col]) {
					lineValue = cellLines[col][lineIdx]
				}
				fmt.Printf(" %-*s |", colWidths[col], lineValue)
			}
			fmt.Println()
		}
	}

	// Print bottom border
	fmt.Print("+")
	for _, width := range colWidths {
		fmt.Print(strings.Repeat("-", width+2))
		fmt.Print("+")
	}
	fmt.Println()
}

File: bin/spice-cli/src/output/table.rs
/*
Copyright 2024-2026 The Spice.ai OSS Authors

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

     https://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

//! Table formatting for CLI output.

use comfy_table::{Cell, Table, presets::NOTHING};
use std::fmt;

/// A trait for types that can be displayed as a table row.
pub trait TableRow {
    /// Returns the column headers for this row type.
    fn headers() -> Vec<&'static str>;

    /// Returns the values for this row as strings.
    fn values(&self) -> Vec<String>;
}

/// Write a table to stdout.
pub fn write_table<T: TableRow>(rows: &[T]) {
    if rows.is_empty() {
        return;
    }

    let mut table = Table::new();
    table.load_preset(NOTHING);

    // Add headers
    let headers: Vec<Cell> = T::headers().into_iter().map(Cell::new).collect();
    table.set_header(headers);

    // Add rows
    for row in rows {
        let cells: Vec<Cell> = row.values().into_iter().map(Cell::new).collect();
        table.add_row(cells);
    }

    println!("{table}");
}

/// A simple table output builder for ad-hoc tables.
pub struct TableOutput {
    headers: Vec<String>,
    rows: Vec<Vec<String>>,
}

impl TableOutput {
    /// Create a new table with the given column headers.
    pub fn new(headers: Vec<&str>) -> Self {
        Self {
            headers: headers.into_iter().map(String::from).collect(),
            rows: Vec::new(),
        }
    }

    /// Add a row to the table.
    pub fn add_row(&mut self, values: Vec<String>) {
        self.rows.push(values);
    }

    /// Print the table to stdout.
    pub fn print(&self) {
        println!("{self}");
    }

    /// Build the comfy-table Table.
    fn build_table(&self) -> Table {
        let mut table = Table::new();
        table.load_preset(NOTHING);

        // Add headers
        let headers: Vec<Cell> = self.headers.iter().map(Cell::new).collect();
        table.set_header(headers);

        // Add rows
        for row in &self.rows {
            let cells: Vec<Cell> = row.iter().map(Cell::new).collect();
            table.add_row(cells);
        }

        table
    }
}

impl fmt::Display for TableOutput {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.build_table())
    }
}
